### PR TITLE
PKG Add Pysam

### PR DESCRIPTION
**Pull request by stevenweaver**
Description

 - adding pysam to packages https://github.com/pysam-developers/pysam
 - This package is heavily used by bioinformaticians who mainly code in python
Checklists

 - Packaging seems to be running smoothly and a success message is shown at the end
 - pysam compiles htslib itself so I decided not to package htslib separately

Ref: https://github.com/pyodide/pyodide/issues/2111

### CODE REVIEW

**Comment by ryanking13** [comment]
The error

```
Error: bad export type for 'bcf_float_missing': undefined
```

means that there are libraries that are not linked correctly (so it has missing symbols). Do you know where that symbol comes from?

**Comment by stevenweaver** [comment]
Yes, it comes from https://github.com/pysam-developers/pysam/blob/cdc0ed12fbe2d7633b8fa47534ab2c2547f66b84/pysam/libchtslib.pxd#L2139 and https://github.com/samtools/htslib/blob/99415e2a2ce26bdbf4e910954330ea769de2c3f0/vcf.c#L84.

In the build.log, it is building a linking:
`ar -rc libhts.a kfunc.o kstring.o bcf_sr_sort.o bgzf.o errmod.o faidx.o header.o hfile.o hts.o hts_expr.o hts_os.o md5.o multipart.o probaln.o realn.o regidx.o region.o sam.o sam_mods.o synced_bcf_reader.o vcf_sweep.o tbx.o textutils.o thread_pool.o vcf.o vcfutils.o cram/cram_codecs.o cram/cram_decode.o cram/cram_encode.o cram/cram_external.o cram/cram_index.o cram/cram_io.o cram/cram_stats.o cram/mFILE.o cram/open_trace_file.o cram/pooled_alloc.o cram/string_alloc.o htscodecs/htscodecs/arith_dynamic.o htscodecs/htscodecs/fqzcomp_qual.o htscodecs/htscodecs/htscodecs.o htscodecs/htscodecs/pack.o htscodecs/htscodecs/rANS_static4x16pr.o htscodecs/htscodecs/rANS_static32x16pr_avx2.o htscodecs/htscodecs/rANS_static32x16pr_avx512.o htscodecs/htscodecs/rANS_static32x16pr_sse4.o htscodecs/htscodecs/rANS_static32x16pr_neon.o htscodecs/htscodecs/rANS_static32x16pr.o htscodecs/htscodecs/rANS_static.o htscodecs/htscodecs/rle.o htscodecs/htscodecs/tokenise_name3.o htscodecs/htscodecs/utils.o  `

Best,
Steven

**Comment by ryanking13** [comment]
Thanks for the info!

I checked the build log and symbols, and `libcbcf.cpython-311-wasm32-emscripten.so` requires that symbol but is not linking to `libhts` for some reason.

Maybe it is a bug of `pysam` (I am not sure, I don't know anything about pysam), but when I look at the setup.py code, I think you can [patch this line](https://github.com/pysam-developers/pysam/blob/cdc0ed12fbe2d7633b8fa47534ab2c2547f66b84/setup.py#L624C15-L627) as follows:

```python
    dict(name="pysam.libcbcf",
         sources=[source_pattern % "bcf"] + os_c_files,
         extra_objects=separate_htslib_objects,
         libraries=libraries_for_pysam_module),
```

to

```python
    dict(name="pysam.libcbcf",
         sources=[source_pattern % "bcf"] + os_c_files,
         extra_objects=htslib_objects,
         libraries=libraries_for_pysam_module),
```

**Comment by stevenweaver** [comment]
Thanks @ryanking13,
I changed the setting `HTSLIB_MODE` setting to `separate`, which appears to have the same affect without writing another patch.

**Comment by stevenweaver** [comment]
Yep, I'll get on it today. Thanks!

**Comment by stevenweaver** [comment]
This will also close #2115

**Comment by stevenweaver** [comment]
Addressed @ryanking13's three requests. Changing PR to "Ready to review".

File: packages/pysam/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    - patches/0001-use-emmake.patch
+    - patches/0001-remove-darwin-check.patch
+requirements:
+  host:
+    - zlib
+build:
+  cflags: |
+    -I$(WASM_LIBRARY_DIR)/include
+    -fPIC
+  cross-script: |
```
Comment (ryanking13) [inline packages/pysam/meta.yaml:20]: I guess these commands inside `cross-script` are for debugging. Could you delete them?

**Review by ryanking13 — APPROVED** [review]
Thanks for the detailed explanation in the patch file! I have one small comment and other than that it looks good.

**Pull request merged**